src/Renci.SshNet/Common/Pipe.cs

```diff
@@ -0,0 +1,47 @@
+﻿namespace Renci.SshNet.Common
+{
+    using System;
+    using System.IO;
+
+    /// <summary>
+    /// A generic pipe to pass through data.
+    /// </summary>
+    internal class Pipe : IDisposable
+    {
+        private readonly LinkedListQueue<byte[]> _queue;
+
+        /// <summary>
+        /// Gets the input stream.
+        /// </summary>
+        /// <value>The input stream.</value>
+        public Stream InputStream { get; private set; }
+
+        /// <summary>
+        /// Gets the output stream.
+        /// </summary>
+        /// <value>The output stream.</value>
+        public Stream OutputStream { get; private set; }
+
+        public Pipe()
+        {
+            _queue = new LinkedListQueue<byte[]>();
+            InputStream = new PipeInputStream(_queue);
```

Comment:
I think we should at least lazily initialize the inputstream.
In that case, we can leave it up to the consumer to dispose it.
In that case, what should PipeInputStream.Read(...) do when the output stream is disposed?
Should it just return 0 signaling EOS?